File: code/packages/ruby/bezier2d/test/test_bezier2d.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "../lib/bezier2d"

class TestQuadraticBezier < Minitest::Test
  include Point2D
  DELTA = 1e-9

  def q
    @q ||= Bezier2D::QuadraticBezier.new(
      Point.new(0, 0), Point.new(1, 2), Point.new(2, 0)
    )
  end

  def test_eval_at_zero
    p = q.eval(0)
    assert_in_delta 0, p.x, DELTA
    assert_in_delta 0, p.y, DELTA
  end

  def test_eval_at_one
    p = q.eval(1)
    assert_in_delta 2, p.x, DELTA
    assert_in_delta 0, p.y, DELTA
  end

  def test_eval_midpoint
    p = q.eval(0.5)
    assert_in_delta 1, p.x, DELTA
    assert_in_delta 1, p.y, DELTA
  end

  def test_split_midpoints
    left, right = q.split(0.5)
    m = q.eval(0.5)
    assert_in_delta m.x, left.p2.x, DELTA
    assert_in_delta m.x, right.p0.x, DELTA
  end

  def test_polyline_straight
    straight = Bezier2D::QuadraticBezier.new(
      Point.new(0, 0), Point.new(1, 0), Point.new(2, 0)
    )
    pts = straight.polyline(0.1)
    assert_equal 2, pts.length
  end

  def test_bbox_contains_endpoints
    bb = q.bbox
    assert bb.x <= 0
    assert bb.x + bb.width >= 2
  end

  def test_elevate_equivalent
    c = q.elevate
    [0, 0.25, 0.5, 0.75, 1].each do |t|
      qp = q.eval(t)
      cp = c.eval(t)
      assert_in_delta qp.x, cp.x, DELTA
      assert_in_delta qp.y, cp.y, DELTA
    end
  end
end

class TestCubicBezier < Minitest::Test
  include Point2D
  DELTA = 1e-9

  def c
    @c ||= Bezier2D::CubicBezier.new(
      Point.new(0, 0), Point.new(1, 2),
      Point.new(3, 2), Point.new(4, 0)
    )
  end

  def test_eval_at_zero
    p = c.eval(0)
    assert_in_delta 0, p.x, DELTA
    assert_in_delta 0, p.y, DELTA
  end

  def test_eval_at_one
    p = c.eval(1)
    assert_in_delta 4, p.x, DELTA
    assert_in_delta 0, p.y, DELTA
  end

  def test_eval_symmetric_midpoint
    # Symmetric cubic: midpoint x should be 2
    p = c.eval(0.5)
    assert_in_delta 2, p.x, DELTA
  end

  def test_split_midpoints
    left, right = c.split(0.5)
    m = c.eval(0.5)
    assert_in_delta m.x, left.p3.x, DELTA
    assert_in_delta m.x, right.p0.x, DELTA
  end

  def test_polyline_straight
    straight = Bezier2D::CubicBezier.new(
      Point.new(0, 0), Point.new(1, 0), Point.new(2, 0), Point.new(3, 0)
    )
    pts = straight.polyline(0.1)
    assert_equal 2, pts.length
  end

  def test_bbox_contains_samples
    bb = c.bbox
    (0..20).each do |i|
      p = c.eval(i / 20.0)
      assert p.x >= bb.x - 1e-6
      assert p.x <= bb.x + bb.width + 1e-6
    end
  end
end

File: code/packages/python/bezier2d/tests/test_bezier2d.py
"""Tests for bezier2d."""
import pytest
from bezier2d import QuadraticBezier, CubicBezier
from point2d import Point

EPS = 1e-9
def approx_eq(a, b): return abs(a - b) < EPS
def pt_eq(a, b): return approx_eq(a.x, b.x) and approx_eq(a.y, b.y)


class TestQuadraticBezier:
    q = QuadraticBezier(Point(0, 0), Point(1, 2), Point(2, 0))

    def test_endpoints(self):
        assert pt_eq(self.q.evaluate(0), self.q.p0)
        assert pt_eq(self.q.evaluate(1), self.q.p2)

    def test_midpoint(self):
        mid = self.q.evaluate(0.5)
        assert approx_eq(mid.x, 1.0) and approx_eq(mid.y, 1.0)

    def test_derivative_at_0(self):
        d = self.q.derivative(0)
        assert approx_eq(d.x, 2.0) and approx_eq(d.y, 4.0)

    def test_split_endpoints(self):
        left, right = self.q.split(0.5)
        split_pt = self.q.evaluate(0.5)
        assert pt_eq(left.p2, split_pt) and pt_eq(right.p0, split_pt)
        assert pt_eq(left.p0, self.q.p0) and pt_eq(right.p2, self.q.p2)

    def test_polyline_straight(self):
        q = QuadraticBezier(Point(0, 0), Point(1, 0), Point(2, 0))
        assert len(q.to_polyline(0.1)) == 2

    def test_polyline_endpoints(self):
        pts = self.q.to_polyline(0.1)
        assert pt_eq(pts[0], self.q.p0) and pt_eq(pts[-1], self.q.p2)

    def test_bounding_box_contains_endpoints(self):
        bb = self.q.bounding_box()
        assert bb.x <= 0 and bb.x + bb.width >= 2

    def test_elevate_equivalent(self):
        c = self.q.elevate()
        for t in [0, 0.25, 0.5, 0.75, 1]:
            assert abs(self.q.evaluate(t).x - c.evaluate(t).x) < 1e-9
            assert abs(self.q.evaluate(t).y - c.evaluate(t).y) < 1e-9


class TestCubicBezier:
    c = CubicBezier(Point(0, 0), Point(1, 2), Point(3, 2), Point(4, 0))

    def test_endpoints(self):
        assert pt_eq(self.c.evaluate(0), self.c.p0)
        assert pt_eq(self.c.evaluate(1), self.c.p3)

    def test_symmetric_midpoint(self):
        assert approx_eq(self.c.evaluate(0.5).x, 2.0)

    def test_derivative_straight(self):
        c = CubicBezier(Point(0, 0), Point(1, 0), Point(2, 0), Point(3, 0))
        d = c.derivative(0)
        assert approx_eq(d.x, 3.0) and approx_eq(d.y, 0.0)

    def test_split_endpoints(self):
        left, right = self.c.split(0.5)
        split_pt = self.c.evaluate(0.5)
        assert pt_eq(left.p3, split_pt) and pt_eq(right.p0, split_pt)

    def test_polyline_straight(self):
        c = CubicBezier(Point(0, 0), Point(1, 0), Point(2, 0), Point(3, 0))
        assert len(c.to_polyline(0.1)) == 2

    def test_polyline_curved(self):
        pts = self.c.to_polyline(0.1)
        assert len(pts) > 2
        assert pt_eq(pts[0], self.c.p0) and pt_eq(pts[-1], self.c.p3)

    def test_bounding_box_contains_samples(self):
        bb = self.c.bounding_box()
        for i in range(21):
            p = self.c.evaluate(i / 20)
            assert p.x >= bb.x - 1e-6 and p.x <= bb.x + bb.width + 1e-6
            assert p.y >= bb.y - 1e-6 and p.y <= bb.y + bb.height + 1e-6
